Correctly handle empty arrays

iteratee opened the pull request:
If the array was of size zero, previous code would output a spurious extra 0 byte.

Also fixes a type problem where decoded Int was returned as Int64. This is a problem for round-trip because the result could not be reserialized.

Commit:
Generics: Correctly serialize empty arrays.

konstantin-azarov commented on src/generic.jl:
```diff
@@ -30,11 +30,14 @@ Writes an array of Avro objects if there is a
     end
-    bytes_written += encode_byte(encoder, zero(UInt8))
+    if l != 0
+        bytes_written += encode_byte(encoder, zero(UInt8))
```
What is this zero for in general and why is 0 length special cased?

iteratee replied:
I've added comments and a link to the spec.

konstantin-azarov replied:
Thanks.

twadleigh commented on src/io.jl:
```diff
@@ -99,8 +99,8 @@ end
-decode_int(decoder::Decoder) = _decode_zigzag(_decode_varint(decoder.stream, Int32))
-decode_long(decoder::Decoder) = _decode_zigzag(_decode_varint(decoder.stream, Int64))
+decode_int(decoder::Decoder)::Int32 = _decode_zigzag(_decode_varint(decoder.stream, Int32))
+decode_long(decoder::Decoder)::Int64 = _decode_zigzag(_decode_varint(decoder.stream, Int64))
```
Did you add these type assertions for the purpose of documentation?

twadleigh replied:
If you actually need the type coersion that this would imply, I would rather tweak the definition of `_decode_varint`, which I think might be the result of the type instability. (Replace the last line with `convert(T, n)`.)

twadleigh replied:
Looking closer at the definition of `_decode_varint`, I can't see how either it or `_decode_zigzag` can fail to return the desired type. Still, I think it would be regarded as suboptimal form to rely on a type coersion on the return value in this way (vs. putting the conversion in the body).

iteratee commented:
On Tue, Feb 4, 2020, 3:47 PM Tracy Wadleigh <notifications@github.com>
wrote:

> *@twadleigh* commented on this pull request.
> ------------------------------
>
> In src/io.jl
> <https://github.com/compound-eye/Avro.jl/pull/2#discussion_r374988566>:
>
> > @@ -99,8 +99,8 @@ end
>  decode_null(::Decoder) = nothing
>  decode_boolean(decoder::Decoder) = read(decoder.stream, Bool)
>
> -decode_int(decoder::Decoder) = _decode_zigzag(_decode_varint(decoder.stream, Int32))
> -decode_long(decoder::Decoder) = _decode_zigzag(_decode_varint(decoder.stream, Int64))
> +decode_int(decoder::Decoder)::Int32 = _decode_zigzag(_decode_varint(decoder.stream, Int32))
> +decode_long(decoder::Decoder)::Int64 = _decode_zigzag(_decode_varint(decoder.stream, Int64))
>
> Did you add these type assertions for the purpose of documentation?
>
No. Without them the deserialization code ends up returning an Int64, which
breaks round tripping.

> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/compound-eye/Avro.jl/pull/2?email_source=notifications&email_token=ABES37GOCK6CWO5GDEVE3WTRBH5AHA5CNFSM4KP7JCJKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCUIDFDQ#pullrequestreview-353383054>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABES37EDDP5NIEXIQECSBOTRBH5AHANCNFSM4KP7JCJA>
> .
>

twadleigh reviewed (approved):
You can merge as is, but I think I might look into how to remove those type assertions.

Commit:
IO: Make _decode_zigzag type stable for round trip stability.

Without type stability, decoding an Int32 could return an Int64, which
would then subsequently fail to serialize.
